File: Rectangle.Windows/src/Rectangle.Windows.WinUI/ViewModels/SnapAreasViewModel.cs
using System.Threading.Tasks;
using Rectangle.Windows.WinUI.Services;

namespace Rectangle.Windows.WinUI.ViewModels
{
    public class SnapAreasViewModel : ObservableObject
    {
        private readonly ConfigService _configService;
        private bool _dragToSnap;
        private bool _restoreSizeOnSnapEnd;
        private bool _snapAnimation;
        private bool _hapticFeedback;
        private int _gapSize;

        public bool DragToSnap
        {
            get => _dragToSnap;
            set { if (SetProperty(ref _dragToSnap, value)) SaveSettings(); }
        }

        public bool RestoreSizeOnSnapEnd
        {
            get => _restoreSizeOnSnapEnd;
            set { if (SetProperty(ref _restoreSizeOnSnapEnd, value)) SaveSettings(); }
        }

        public bool SnapAnimation
        {
            get => _snapAnimation;
            set { if (SetProperty(ref _snapAnimation, value)) SaveSettings(); }
        }

        public bool HapticFeedback
        {
            get => _hapticFeedback;
            set { if (SetProperty(ref _hapticFeedback, value)) SaveSettings(); }
        }

        public int GapSize
        {
            get => _gapSize;
            set { if (SetProperty(ref _gapSize, value)) SaveSettings(); }
        }

        public SnapAreasViewModel()
        {
            _configService = new ConfigService();
        }

        public async Task LoadSettingsAsync()
        {
            await Task.Run(() =>
            {
                var config = _configService.Load();
                _dragToSnap = config.SnapAreas?.DragToSnap ?? true;
                _restoreSizeOnSnapEnd = config.SnapAreas?.RestoreSizeOnSnapEnd ?? true;
                _snapAnimation = config.SnapAreas?.SnapAnimation ?? false;
                _hapticFeedback = config.SnapAreas?.HapticFeedback ?? false;
                _gapSize = config.GapSize;
                OnPropertyChanged(nameof(DragToSnap));
                OnPropertyChanged(nameof(RestoreSizeOnSnapEnd));
                OnPropertyChanged(nameof(SnapAnimation));
                OnPropertyChanged(nameof(HapticFeedback));
                OnPropertyChanged(nameof(GapSize));
            });
        }

        private void SaveSettings()
        {
            var config = _configService.Load();
            config.SnapAreas ??= new SnapAreaConfig();
            config.SnapAreas.DragToSnap = _dragToSnap;
            config.SnapAreas.RestoreSizeOnSnapEnd = _restoreSizeOnSnapEnd;
            config.SnapAreas.SnapAnimation = _snapAnimation;
            config.SnapAreas.HapticFeedback = _hapticFeedback;
            config.GapSize = _gapSize;
            _configService.Save(config);
        }
    }
}

File: Rectangle.Windows/src/Rectangle.Windows.WinUI.Cpp/Views/SnapAreasViewModel.cpp
#include "pch.h"
#include "Views/SnapAreasViewModel.h"
#include "Services/Logger.h"
#include "Services/ConfigService.h"

namespace winrt::Rectangle::Views
{
    SnapAreasViewModel::SnapAreasViewModel()
    {
        Logger::Instance().Info(L"SnapAreasViewModel", L"SnapAreasViewModel constructed");
    }

    void SnapAreasViewModel::LoadSettings()
    {
        Logger::Instance().Info(L"SnapAreasViewModel", L"Loading settings");

        Services::ConfigService configService;
        auto config = configService.Load();

        m_dragToSnap = config.SnapAreas.DragToSnap;
        m_restoreSizeOnSnapEnd = config.SnapAreas.RestoreSizeOnSnapEnd;
        m_snapAnimation = config.SnapAreas.SnapAnimation;
        m_hapticFeedback = config.SnapAreas.HapticFeedback;
        m_gapSize = config.GapSize;

        if (m_onChanged)
        {
            m_onChanged();
        }
    }

    void SnapAreasViewModel::SaveSettings()
    {
        Logger::Instance().Info(L"SnapAreasViewModel", L"Saving settings");

        Services::ConfigService configService;
        auto config = configService.Load();

        config.SnapAreas.DragToSnap = m_dragToSnap;
        config.SnapAreas.RestoreSizeOnSnapEnd = m_restoreSizeOnSnapEnd;
        config.SnapAreas.SnapAnimation = m_snapAnimation;
        config.SnapAreas.HapticFeedback = m_hapticFeedback;
        config.GapSize = m_gapSize;

        configService.Save(config);
    }

    void SnapAreasViewModel::SetDragToSnap(bool value)
    {
        if (m_dragToSnap != value)
        {
            m_dragToSnap = value;
            SaveSettings();
        }
    }

    void SnapAreasViewModel::SetRestoreSizeOnSnapEnd(bool value)
    {
        if (m_restoreSizeOnSnapEnd != value)
        {
            m_restoreSizeOnSnapEnd = value;
            SaveSettings();
        }
    }

    void SnapAreasViewModel::SetSnapAnimation(bool value)
    {
        if (m_snapAnimation != value)
        {
            m_snapAnimation = value;
            SaveSettings();
        }
    }

    void SnapAreasViewModel::SetHapticFeedback(bool value)
    {
        if (m_hapticFeedback != value)
        {
            m_hapticFeedback = value;
            SaveSettings();
        }
    }

    void SnapAreasViewModel::SetGapSize(int32_t value)
    {
        if (m_gapSize != value)
        {
            m_gapSize = value;
            SaveSettings();
        }
    }
}